lib/spack/spack/detection/path.py

```diff
@@ -241,15 +241,19 @@ def by_executable(
         path_hints: list of paths to be searched. If None the list will be
             constructed based on the PATH environment variable.
     """
+    no_path_hints = path_hints is None
```

Comment:
An empty list should also mean no path hints so I think `no_path_hints = not bool(path_hints)` would be better.

Reply:
`path_hints is None` was a deliberate choice to avoid casting `path_hints` to bool as this mimic the behavior of the next line, which also uses `None` as opposed to booleans, and in doing so, ensures behavior dependent on `no_path_hints` remains the same as if it were part of the following command, which was the intent behind `no_path_hints`.
I assumed the choice below to explicitly use `None` was a deliberate one. If this is not the case, then I'll refactor both. Which ever way we decide to go, these two lines should use the same check. I only don't refer to `no_path_hints` in the following line because `mypy` can't seem to handle it and I've already spent more time than I'd like trying to reconcile that.

Reply:
I believe the intent is to register if the user passed anything as argument, instead of using the default value. In that sense:
```
path_hints is None
```
is the correct choice. In particular in the line below, this allows distinguishing among "default value" (for which we look into the PATH env variable), and an empty list that was passed explicitly.